corgi/tasks/managed_services.py

```diff
@@ -59,42 +79,60 @@ def cpu_manifest_service(product_stream_id: str, service_components: list) -> No
-                name=service_component["name"],
-                build_type=SoftwareBuild.Type.APP_INTERFACE,
-                meta_attr__service=service.name,
-            )
-            for build in past_builds:
-                ProductComponentRelation.objects.filter(software_build=build).delete()
-                build.delete()
-
             build_id = now.strftime("%Y%m%d%H%M%S%f")
-            build = SoftwareBuild.objects.create(
-                name=service_component["name"],
-                build_type=SoftwareBuild.Type.APP_INTERFACE,
-                build_id=build_id,
-                completion_time=now,
-                meta_attr={"service": service.name},
-            )
-
-            root_component = Component.objects.create(
+            try:
+                # get_or_create requires all keyword arguments to appear in
+                # a unique_together constraint, like build_id and build_type
+                # Otherwise it will not be atomic and can behave incorrectly
+                # Since we don't have a database index for name + build_type,
+                # we try to find an APP_INTERFACE build with a matching name first
+                # And update the list of services it belongs to
+                # If that fails, we'll create a new build
+                # So Corgi reuses a component when services reuse the component
+                build = SoftwareBuild.objects.get(
+                    name=service_component["name"],
+                    build_type=SoftwareBuild.Type.APP_INTERFACE,
+                )
+                build.meta_attr["services"].append(service.name)
+                build.save()
+            except SoftwareBuild.DoesNotExist:
+                build = SoftwareBuild.objects.create(
+                    name=service_component["name"],
+                    build_type=SoftwareBuild.Type.APP_INTERFACE,
+                    build_id=build_id,
+                    completion_time=now,
+                    meta_attr={"services": [service.name]},
+                )
+
+            # Root components can only be linked to one build
+            # if we use the same build / component for two different services,
+            # we will analyze the same component twice
+            # this should be OK and the taxonomy should be the same.
+            # Or if not, merging the data is probably what we want.
+
+            # Use only indexed / unique_together kwargs to find the object
+            root_component, _ = Component.objects.update_or_create(
                 type=Component.Type.CONTAINER_IMAGE if quay_repo else Component.Type.GITHUB,
                 name=service_component["name"],
                 version=component_version,
                 release="",
                 arch="noarch",
-                namespace=Component.Namespace.REDHAT,
-                software_build=build,
+                # TODO: What if component exists but isn't in REDHAT namespace / linked to a build?
```

Comment:
The root component here is either a Red Hat-built image or a Red Hat-owned Git repo, so I think it should never be anything else than a Red Hat-namespaced component, no?